File: examples/rust_lib/android/src/main/java/com/sergioribera/rustlib/MainActivity.java
package com.sergioribera.rustlib;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity implements JNICallback {

    private static final String TAG = "MainActivity";

    // Used to load the 'rust' library on application startup.
    static {
        System.loadLibrary("rust_lib");
    }

    TextView textView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        textView = (TextView) findViewById(R.id.sample_text);

        invokeCallbackViaJNI(this);
    }

    /**
     * A native method that is implemented by the 'rust' native library,
     * which is packaged with this application.
     */
    public static native void invokeCallbackViaJNI(JNICallback callback);

    @Override
    public void callback(String string) {
        textView.append("From JNI: " + string + "\n");
    }
}

File: examples/rust_lib/android/src/main/java/com/sergioribera/rustlib/MainActivity.kt
package com.sergioribera.rustlib

import android.os.Bundle
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.sergioribera.rustlib.R

class MainActivity : AppCompatActivity(), JNICallback {
    var textView: TextView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        textView = findViewById(R.id.sample_text)

        invokeCallbackViaJNI(this)
    }

    override fun callback(string: String?) {
        textView!!.append("From JNI: $string\n")
    }

    /**
     * A native method that is implemented by the 'rust' native library,
     * which is packaged with this application.
     */
    external fun invokeCallbackViaJNI(callback: JNICallback?)

    companion object {
        // Used to load the 'rust' library on application startup.
        init {
            System.loadLibrary("rust_lib")
        }
    }
}
